### PR TITLE
semaphore: change semaphore count to atomic operation

### PR DESCRIPTION
**Pull request by zyfeier**
## Summary

Add sem_wait fast operations, use atomic to ensure atomicity of semcount operations, and do not depend on critical section.

Test with robot board:
before modify:
nxmutex_lock cost: 78 ns
nxmutex_unlock cost: 82 ns

after modify:
nxmutex_lock cost: 28 ns
nxmutex_unlock cost: 14 ns

## Impact

semaphore

## Testing

bes board with monkey test pass
sabre-6quad:smp with ostest pass
sim:citest pass

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



No.  The PR description lacks several key details required by the NuttX guidelines.

Here's a breakdown:

* **Insufficient Summary:** While the summary mentions *what* changed, it lacks crucial *why* information.  Why was this change necessary? What problem does it solve? What are the benefits of using atomic operations and avoiding the critical section?
* **Missing Impact Details:**  Simply stating "semaphore" is insufficient.  The impact section needs specific YES/NO answers for *all* listed questions.  For any "YES" answers, a detailed explanation is required. For example, even if the user impact is NO, explicitly state "Impact on user: NO."  Consider these points for your impact assessment:
    * **Performance:**  The summary mentions performance improvements.  This should be explicitly stated in the Impact section, quantifying the improvement.
    * **Behavior Change:** Does this change the behavior of semaphores in any way beyond performance?  Even subtle changes can affect existing applications.
    * **Portability:**  Does this change affect portability to other architectures? Does it rely on specific atomic operations that might not be available everywhere?
* **Incomplete Testing Information:** While testing environments are mentioned, it lacks details required by the template.  The provided information is too vague.
    * **Build Host:** Specify OS version, compiler version (e.g., GCC 12.2), etc.
    * **Target:** Be specific. Instead of "sim," specify the simulator being used (e.g., qemu-system-arm).  For boards, specify the configuration used (e.g., `sabre-6quad:smp`).
* **Missing "Before" and "After" Logs:** The provided timing information is helpful, but insufficient.  The template specifically requests *logs*, not just summarized timings. Show relevant log excerpts demonstrating the semaphore behavior before and after the change.  Include more than just the timings - show the context of the semaphore usage in the test.
* **Missing Issue References:** Are there any related NuttX or NuttX Apps issues that prompted this change? If so, include links.


**In short:** The PR needs to be significantly expanded to address these missing details before it can be considered complete according to the NuttX requirements.  Focus on providing specific, detailed answers to all questions in the template.

File: sched/semaphore/sem_wait.c [hunk 1]
```diff
@@ -102,7 +96,6 @@ int nxsem_wait(FAR sem_t *sem)
           return ret;
         }
 
-      sem->semcount--;
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_wait.c:105]: need restore count at line 95

File: sched/semaphore/sem_trywait.c [hunk 2]
```diff
@@ -68,49 +138,28 @@
...
-
-  if (sem->semcount > 0)
+#if !defined(CONFIG_PRIORITY_INHERITANCE) && !defined(CONFIG_PRIORITY_PROTECT)
+  if (sem->flags & SEM_TYPE_MUTEX)
     {
-      /* It is, let the task take the semaphore */
-
-      ret = nxsem_protect_wait(sem);
-      if (ret < 0)
+      int32_t old = 1;
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_trywait.c:154]: All similar place change int32_t to atomic_t

File: sched/semaphore/sem_wait.c [hunk 3]
```diff
@@ -224,6 +213,63 @@ int nxsem_wait(FAR sem_t *sem)
...
+  DEBUGASSERT(!OSINIT_IDLELOOP() || !sched_idletask());
+
+  /* If this is a mutex, we can try to get the mutex in fast mode,
+   * else try to get it in slow mode.
+   */
+
+#if !defined(CONFIG_PRIORITY_INHERITANCE) && !defined(CONFIG_PRIORITY_PROTECT)
+  if (sem->flags & SEM_TYPE_MUTEX)
+    {
+      int32_t old = 1;
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_wait.c:262]: atomic_t

File: sched/semaphore/sem_trywait.c [hunk 4]
```diff
@@ -38,6 +38,76 @@
...
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+static int nxsem_trywait_slow(FAR sem_t *sem)
+{
+  FAR struct tcb_s *rtcb;
+  irqstate_t flags;
+  int32_t semcount;
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_trywait.c:67]: atomic_t

File: sched/semaphore/sem_trywait.c [hunk 5]
```diff
@@ -38,6 +38,76 @@
...
+   */
+
+  flags = enter_critical_section();
+  rtcb = this_task();
+
+  /* If the semaphore is available, give it to the requesting task */
+
+  do
+    {
+      semcount = atomic_read(NXSEM_COUNT(sem));
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_trywait.c:81]: move before line 79

File: sched/semaphore/sem_reset.c [hunk 6]
```diff
@@ -60,6 +60,7 @@
 int nxsem_reset(FAR sem_t *sem, int16_t count)
 {
   irqstate_t flags;
+  int32_t semcount;
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_reset.c:63]: atomic_t

File: sched/semaphore/sem_reset.c [hunk 7]
```diff
@@ -98,10 +99,15 @@ int nxsem_reset(FAR sem_t *sem, int16_t count)
    * value of sem->semcount is already correct in this case.
    */
 
-  if (sem->semcount >= 0)
+  do
     {
-      sem->semcount = count;
+      semcount = atomic_read(NXSEM_COUNT(sem));
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_reset.c:104]: move before line 101

File: sched/semaphore/sem_post.c [hunk 8]
```diff
@@ -217,3 +214,58 @@ int nxsem_post(FAR sem_t *sem)
...
+  DEBUGASSERT(sem != NULL);
+
+  /* If this is a mutex, we can try to unlock the mutex in fast mode,
+   * else try to get it in slow mode.
+   */
+
+#if !defined(CONFIG_PRIORITY_INHERITANCE) && !defined(CONFIG_PRIORITY_PROTECT)
+  if (sem->flags & SEM_TYPE_MUTEX)
+    {
+      int32_t old = 0;
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_post.c:262]: atomic_t

File: sched/semaphore/sem_post.c [hunk 9]
```diff
@@ -78,21 +78,20 @@ int nxsem_post(FAR sem_t *sem)
...
    */
 
   flags = enter_critical_section();
 
-  sem_count = sem->semcount;
+  sem_count = atomic_fetch_add(NXSEM_COUNT(sem), 1);
 
   /* Check the maximum allowable value */
 
   if (sem_count >= SEM_VALUE_MAX)
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_post.c:92]: SEM_VALUE_MAX could change to big value

File: sched/semaphore/sem_destroy.c [hunk 10]
```diff
@@ -60,6 +60,8 @@
 
 int nxsem_destroy(FAR sem_t *sem)
 {
+  int32_t old;
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_destroy.c:63]: atomic_t

File: sched/semaphore/sem_destroy.c [hunk 11]
```diff
@@ -72,10 +74,15 @@ int nxsem_destroy(FAR sem_t *sem)
    * leave the count unchanged but still return OK.
    */
 
-  if (sem->semcount >= 0)
+  do
     {
-      sem->semcount = 1;
+      old = atomic_read(NXSEM_COUNT(sem));
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_destroy.c:79]: move before line 75

File: sched/semaphore/sem_trywait.c [hunk 12]
```diff
@@ -38,6 +38,76 @@
...
+  if (ret < 0)
+    {
+      atomic_fetch_add(NXSEM_COUNT(sem), 1);
+      leave_critical_section(flags);
+      return ret;
+    }
+
+  nxsem_add_holder(sem);
+  rtcb->waitobj = NULL;
+  ret = OK;
```
Comment (xiaoxiang781216) [inline sched/semaphore/sem_trywait.c:103]: remove

File: tools/ci/testrun/script/test_open_posix/test_openposix_.py [hunk 13]
```diff
@@ -6984,7 +6984,7 @@ def test_ltp_interfaces_pthread_rwlock_timedrdlock_5_1(p):
 def test_ltp_interfaces_sem_init_6_1(p):
     ret = p.sendCommand(
         "ltp_interfaces_sem_init_6_1",
-        ["PASSED", "passed", "Passed", "PASS"],
+        ["PASSED", "passed", "Passed", "PASS", "skipped"],
```
Comment (xiaoxiang781216) [inline tools/ci/testrun/script/test_open_posix/test_openposix_.py:6987]: why skip

Comment (zyfeier) [reply tools/ci/testrun/script/test_open_posix/test_openposix_.py:6987]: if SEM_VALUE_MAX >= INT_MAX, this case will skipped.
int main(void)
{
	sem_t mysemp;
	int counter = SEM_VALUE_MAX;

	if (SEM_VALUE_MAX >= INT_MAX) {
		puts("Test skipped");
		return PTS_PASS;
	}

	++counter;
	sem_init(&mysemp, 0, counter);

	if (errno == EINVAL) {
		puts("TEST PASSED");
		return PTS_PASS;
	} else {
		puts("TEST FAILED");
		return PTS_FAIL;
	}
}

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**